### PR TITLE
Update the Katib design

### PR DESCRIPTION
**Pull request by wangkuiyi**

### CODE REVIEW

File: doc/design/katib_on_sqlflow.md [hunk 1]
```diff
@@ -1,70 +1,91 @@
...
-    - `validation_dataset` is required in Katib jobs. This indicates the data used for testing in model training in Katib.  
-- `my_xgb_model` is the name of the trained model. 
 
+### Deep Learning Models
+
+For deep learning models, the case is complicated.  Each model has its own set of hyperparameters, and their ranges might vary significantly.  Our proposed solution is to utilize the [model zoo](model_zoo.md).  In particular, users might train a model defined in the zoo with various datasets, in various experiments, with manually specified hyperparameters.  After the training, some users might publish their trained models, including the estimated parameters and the specified hyperparameters.  Given these published hyperparameter values, the [Bayesian hyperparameter optimization](https://en.wikipedia.org/wiki/Hyperparameter_optimization#Bayesian_optimization)  for hyperparameter tuning.  We are working on such a Bayesian approach that doesn't require explicit specification of hyperparameter ranges.  We plan to contribute it to Katib.
+
+### Tigger Hyperparameter Tunning
```
Comment (shendiaomo) [inline doc/design/katib_on_sqlflow.md]: `s/Tigger/Trigger` and `s/Tunning/Tuning`?

File: doc/design/katib_on_sqlflow.md [hunk 2]
```diff
@@ -1,70 +1,91 @@
...
+SELECT * FROM a, b WHERE a.id = b.id INTO c;
+SELECT * FROM c TO TRAIN model_def WITH objective=multi:softmax, eta=1 LABEL class INTO my_xgb_model;
+```
+
+The `codegen_couler.go` might generate the following Couler program.
+
+```python
+couler.maxcompute.run("""SELECT * FROM a, b WHERE a.id = b.id INTO c;
+                         SELECT * FROM c INTO temp""")
+couler.maxcompute.export(table="temp", file="/hdfs/temp")
```
Comment (shendiaomo) [inline doc/design/katib_on_sqlflow.md:81]: It seems this `temp` is unnecessary.

Comment (wangkuiyi) [reply doc/design/katib_on_sqlflow.md:81]: Oh. I mean to export the table temp to be a file on HDFS named temp. So, I think we need both "temp", am I right?

Comment (shendiaomo) [reply doc/design/katib_on_sqlflow.md:81]: Could we export the table `c` to the HDFS file？It seems `temp` is identical to `c`?

File: doc/design/katib_on_sqlflow.md [hunk 3]
```diff
@@ -1,70 +1,91 @@
...
+couler.katib.train(model_def, data="/hdfs/temp")
+```
+
+## `couler.katib.train(...)`
+
+Considering Katib itself supports multiple models and frameworks, and more may come in the future, we introduce the following Couler function.
+
+```python
+def couler.katib.train(model_def=None, hyperparameters={})
```
Comment (Yancey0623) [inline doc/design/katib_on_sqlflow.md:90]: Can you show more details about the khatib train step, such as:
1. where does the model_def come from in SQLFlow?
1. In this example, `hyperparameters ` is empty, what and how does this argument to use in SQLFlow?
1. Do we need `couler.katib.predict` ?

Comment (samplise) [reply doc/design/katib_on_sqlflow.md:90]: Sure.

1.1 As @wangkuiyi suggested, programmers need not specify to use katib explicitly. Instead, sqlflow will check if programmers have input all necessary model parameters. If it is, sqlflow invokes submitter to train model (e.g., xgboost submiitter); otherwise, sqlflow uses katib via couler.
1.2  As the xgboost example above, the model_def can be something like `xgboost:gbtree`.
2. `hyperparameters={"booster": "gbtree",  "objective": "multi:softmax", "eta": 1.0, "max_round": [20, 100], "max_depth: []" }`.
3. No. Katib is only used for model tuning.

File: doc/design/katib_on_sqlflow.md [hunk 4]
```diff
@@ -1,70 +1,91 @@
...
-SELECT * FROM train_table
-TO TRAIN a_data_scientist/xgboost:v0.5/katib.xgboost.gbtree
+- `max_depth` in the range [2,10], and
+- `num_round` in the range [50, 100].
+
+With the introduction of auto hyperparameter tunning, we hope that users don't need to specify the `num_round` and `max_depth` values in the following SQL statement.
+
+```sql
+SELECT * FROM a_dataset_table
+TO TRAIN a_data_scientist/xgboost_models:v0.5/a_gbtree_model
```
Comment (samplise) [inline doc/design/katib_on_sqlflow.md:30]: I realize there are potential problems to allow programmers to use their own image in katib tuning job.
1. The collector in Katib requires the job container's output to follow a certain format, like `accuracy = 0.97`. And users need to tell Katib to collect data for `accuracy` in katib job yaml.
2. Katib does not obtain the tuned model. When one pod finishes tuning, it needs to write the tuned model to a store, e.g., odps, then the following pred can make use of it.

I think it requires too much on programmers in their own image. Maybe sqlflow invokes submitters to train model when programmers specify the image; otherwise, sqlflow uses katib.

File: doc/design/katib_on_sqlflow.md (hunk 2 above)
Comment (wangkuiyi) [reply doc/design/katib_on_sqlflow.md:81]: "INTO temp" means saving the output from the SELECT operation into a new table named temp.

**Review by samplise — APPROVED** [review]
